### PR TITLE
Menu Support and AXIS Targets

### PR DESCRIPTION
**Pull request by brandonrc**
## Changes
Added AXIS TX target along with menu system for TFT and OLED
### AXIS TX
- G Sensor (Motion Mode)
- Temperature Monitor 
- Dynamic Fan
- TFT Support
- Power Calibration

### MENU (ALL TX's equipped with OLED or TFT)
Note: If a TX is equipped with a display but no joystick you will still get a status page that updates when the settings change. 
- Supports OLED and TFT
- Pages
- Joystick support (5 way-button)
- Custom LOGOs

### CODE REVIEW

File: src/include/target/NamimnoRC_FLASH_2400_OLED_TX.h [hunk 1]
```diff
@@ -40,9 +41,13 @@
...
 #define GPIO_PIN_LED_WS2812     4
 
-// Output Power
+/* Output Power */
 #define MinPower                PWR_25mW
 #define MaxPower                PWR_1000mW
 #define POWER_OUTPUT_VALUES     {-18,-13,-10,-5,-2,3}
 
+/* Joystick values              {UP, DOWN, LEFT, RIGHT, ENTER, IDLE}*/
+#define JOY_ADC_VALUES          {1850, 900, 490, 1427, 0, 2978}
```
Comment (brandonrc) [inline src/include/target/NamimnoRC_FLASH_2400_OLED_TX.h:50]: @FOG-Yamato are these the correct values. I do not have a NamimnoRC

File: src/include/target/NamimnoRC_VOYAGER_900_OLED_TX.h [hunk 2]
```diff
@@ -43,4 +43,7 @@
 #define POWER_OUTPUT_DACWRITE
 #define MinPower                PWR_25mW
 #define MaxPower                PWR_2000mW
-#define POWER_OUTPUT_VALUES     {57,67,76,90,112,132,167,201}
\ No newline at end of file
+#define POWER_OUTPUT_VALUES     {57,67,76,90,112,132,167,201}
+
+/* Joystick values              {UP, DOWN, LEFT, RIGHT, ENTER, IDLE}*/
+#define JOY_ADC_VALUES          {1850, 900, 490, 1427, 0, 2978}
```
Comment (brandonrc) [inline src/include/target/NamimnoRC_VOYAGER_900_OLED_TX.h:49]: @FOG-Yamato same here just like the 2400 target. 

Are these the right ADC Values??

File: src/lib/CONFIG/config.cpp [hunk 3]
```diff
@@ -302,6 +341,8 @@ TxConfig::SetDefaults()
         SetBoostChannel(0);
         SetSwitchMode((uint8_t)smHybrid);
         SetModelMatch(false);
+        SetFanMode(0);
+        SetMotionMode(0);
```
Comment (brandonrc) [inline src/lib/CONFIG/config.cpp:345]: Should these be protected by `#ifdef HAS_GSENSOR`??

Comment (pkendall64) [reply src/lib/CONFIG/config.cpp:345]: I think we should leave the function and and guard the body so we don't have guards all over the place.

File: src/include/target/NamimnoRC_FLASH_2400_OLED_TX.h (hunk 1 above)
Comment (FOG-Yamato) [reply src/include/target/NamimnoRC_FLASH_2400_OLED_TX.h:50]: Yes these should be the values necessary.

File: src/include/target/NamimnoRC_VOYAGER_900_OLED_TX.h (hunk 2 above)
Comment (FOG-Yamato) [reply src/include/target/NamimnoRC_VOYAGER_900_OLED_TX.h:49]: Same as the 2400 target.

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+ * Must be placed after the software reset
+ *
+ */
+void STK8xxx::STK8xxx_Anymotion_init()
+{
+	unsigned char ARegAddr, ARegWriteValue;
+	
+	/* Enable X Y Z-axis any-motion (slope) interrupt */
+    ARegAddr       = STK8xxx_REG_INTEN1;
+    ARegWriteValue = 0x07;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:38]: Replace with

#define STK8xxx_REG_INTEN1_EN 0x07

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+ * Must be placed after the software reset
+ *
+ */
+void STK8xxx::STK8xxx_Sigmotion_init()
+{
+	unsigned char SRegAddr, SRegWriteValue;
+	
+	/* Enable X Y Z-axis sig-motion (slope) interrupt */
+    SRegAddr       = STK8xxx_REG_INTEN1;
+    SRegWriteValue = 0x07;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:69]: #define STK8xxx_REG_INTEN1_EN 0x07

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+ * Must be placed after the software reset
+ *
+ */
+void STK8xxx::STK8xxx_Disable_Motion()
+{
+	unsigned char ARegAddr, ARegWriteValue;
+	
+	/* Disable X Y Z-axis motion (slope) interrupt */
+    ARegAddr       = STK8xxx_REG_INTEN1;
+    ARegWriteValue = 0x00;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:100]: #define STK8xxx_REG_INTEN1_DISABLE 0x00

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+	unsigned char ARegAddr, ARegWriteValue;
+	
+	/* Enable X Y Z-axis any-motion (slope) interrupt */
+    ARegAddr       = STK8xxx_REG_INTEN1;
+    ARegWriteValue = 0x07;
+    WriteAccRegister(ARegAddr, ARegWriteValue);
+
+	/* Set anymotion Interrupt trigger threshold */
+    ARegAddr       = STK8xxx_REG_SLOPETHD;
+    ARegWriteValue = 0x14;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:43]: This is the default value and can be removed.  Or replaced with

#define STK8xxx_REG_SLOPETHD_DEFAULT 0x14

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+	unsigned char SRegAddr, SRegWriteValue;
+	
+	/* Enable X Y Z-axis sig-motion (slope) interrupt */
+    SRegAddr       = STK8xxx_REG_INTEN1;
+    SRegWriteValue = 0x07;
+    WriteAccRegister(SRegAddr, SRegWriteValue);
+
+	/* Set sig-motion Interrupt trigger threshold */
+    SRegAddr       = STK8xxx_REG_SLOPETHD;
+    SRegWriteValue = 0x14;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:74]: This is the default value and can be removed.  Or replaced with

#define STK8xxx_REG_SLOPETHD_DEFAULT 0x14

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    WriteAccRegister(ARegAddr, ARegWriteValue);
+
+	/* Enable any-motion */
+    ARegAddr       = STK8xxx_REG_SIGMOT2;
+    ARegWriteValue = 0x04; 
+    WriteAccRegister(ARegAddr, ARegWriteValue);
+
+	/* Map any-motion (slope) interrupt to INT1 */
+    ARegAddr       = STK8xxx_REG_INTMAP1;
+    ARegWriteValue = 0x04;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:53]: #define STK8xxx_REG_INTMAP1_ANYMOT2INT1 0x04

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    WriteAccRegister(SRegAddr, SRegWriteValue);
+
+	/* Enable significant motion */
+    SRegAddr       = STK8xxx_REG_SIGMOT2;
+    SRegWriteValue = 0x02; 
+    WriteAccRegister(SRegAddr, SRegWriteValue);
+
+	/* Map significant motion interrupt to INT1 */
+    SRegAddr       = STK8xxx_REG_INTMAP1;
+    SRegWriteValue = 0x01;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:84]: #define STK8xxx_REG_INTMAP1_SIGMOT2INT1 0x01

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+    WriteAccRegister(ARegAddr, ARegWriteValue);
+
+	/* Set anymotion Interrupt trigger threshold */
+    ARegAddr       = STK8xxx_REG_SLOPETHD;
+    ARegWriteValue = 0x14; 
+    WriteAccRegister(ARegAddr, ARegWriteValue);
+
+	/* Enable any-motion */
+    ARegAddr       = STK8xxx_REG_SIGMOT2;
+    ARegWriteValue = 0x04;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:48]: #define STK8xxx_REG_SIGMOT2_ANY_MOT_EN 0x04

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+    WriteAccRegister(SRegAddr, SRegWriteValue);
+
+	/* Set sig-motion Interrupt trigger threshold */
+    SRegAddr       = STK8xxx_REG_SLOPETHD;
+    SRegWriteValue = 0x14; 
+    WriteAccRegister(SRegAddr, SRegWriteValue);
+
+	/* Enable significant motion */
+    SRegAddr       = STK8xxx_REG_SIGMOT2;
+    SRegWriteValue = 0x02;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:79]: #define STK8xxx_REG_SIGMOT2_SIG_MOT_EN 0x04

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+    unsigned char RegAddr, RegWriteValue;
+
+	if(!STK8xxx_Check_chipid())
+	{
+		return -1;
+    }
+		
+    /* soft-reset */
+	RegAddr       = STK8xxx_REG_SWRST;
+    RegWriteValue = 0xB6;
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:156]: #define STK8xxx_REG_SWRST_RESET_ALL_REGISTERS 0xB6

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+	delay(50);//unit ms
+		
+    /* set range, resolution */
+    RegAddr       = STK8xxx_REG_RANGESEL;
+    RegWriteValue = STK8xxx_RANGE_4G; // range = +/-4g
+    WriteAccRegister(RegAddr, RegWriteValue);
+
+	/* set power mode */
+	RegAddr 	  = STK8xxx_REG_POWMODE;
+	RegWriteValue = 0x00;	// active mode
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:167]: #define STK8xxx_REG_POWMODE_DISABLE_ALL 0x00

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+	WriteAccRegister(RegAddr, RegWriteValue);
+	
+	/* set bandwidth */
+    RegAddr       = STK8xxx_REG_BWSEL;
+    RegWriteValue = 0x0F; // bandwidth = 1000Hz
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:172]: #define STK8xxx_REG_BWSEL_1000Hz 0x0F

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+    WriteAccRegister(RegAddr, RegWriteValue);
+	
+	//STK8xxx_Anymotion_init();
+	//STK8xxx_Sigmotion_init();
+
+    /* set i2c watch dog */
+    RegAddr       = STK8xxx_REG_INTFCFG;
+    RegWriteValue = 0x04; // enable watch dog
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:180]: #define STK8xxx_REG_INTFCFG_I2C_WDT_EN 0x04

File: src/lib/GSENSOR/stk8baxx.cpp [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+    WriteAccRegister(RegAddr, RegWriteValue);
+
+    /* int config */
+    RegAddr       = STK8xxx_REG_INTCFG1;
+    RegWriteValue = 0x05; // INT1/INT2 push-pull, active high
```
Comment (JyeSmith) [inline src/lib/GSENSOR/stk8baxx.cpp:185]: Should this be 0x01 @jonesfu ?

image

Comment (jonesfu) [reply src/lib/GSENSOR/stk8baxx.cpp]: Yes, the value should be 0x01.  This code is inherited from old chip.

Comment (pkendall64) [reply src/lib/GSENSOR/stk8baxx.cpp]: Awesome, thank you.

File: src/lib/SCREEN/TFT/tftscreen.cpp [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+
+    tft.setSwapBytes(true);
+
+    tft.pushImage(0, 0, INIT_PAGE_LOGO_X, INIT_PAGE_LOGO_Y, axis_logo);
+
+    tft.fillRect(SCREEN_FONT_GAP, INIT_PAGE_FONT_START_Y - INIT_PAGE_FONT_PADDING,
+                    SCREEN_X - SCREEN_FONT_GAP*2, SCREEN_NORMAL_FONT_SIZE + INIT_PAGE_FONT_PADDING*2, TFT_BLACK);
+
+    char buffer[50];
+    sprintf(buffer, "THOR-%s  ELRS-", STRING_THOR_VERSION);
```
Comment (JyeSmith) [inline src/lib/SCREEN/TFT/tftscreen.cpp:113]: This isnt generic enough if another manufacture wants to use a TFT.

**Review by JyeSmith — CHANGES_REQUESTED** [review]

**Review by JyeSmith — APPROVED** [review]

**Review by pkendall64 — APPROVED** [review]

**Pull request merged**